### PR TITLE
NO-ISSUE: test strategy

### PR DESCRIPTION
**Pull request by mangelajo**

### CODE REVIEW

**Review by hexfusion — COMMENTED** [review]
Just my 2c but it feels a little early to become so opinionated on testing. The goal should be to add more tests however they look. Then over time, these opinions will work themselves out?

I see these short term goals
- add e2e
- extend unit tests + integration in the code base for now?
- add micro benchmarks as unit tests

Since this doc will evolve overtime could we simplify this to two sections.

- unit/integration
- e2e

**Comment by mangelajo** [comment]
> Just my 2c but it feels a little early to become so opinionated on testing. The goal should be to add more tests however they look. Then over time, these opinions will work themselves out?
> 
> I see these short term goals
> 
> * add e2e
> * extend unit tests + integration in the code base for now?
> * add micro benchmarks as unit tests
> 
> Since this doc will evolve overtime could we simplify this to two sections.
> 
> * unit/integration
> * e2e


Does it really matter where the tests live?, it's actually not a difference for us I think. I just want to make sure we can have a very clear view of the tests we have, so we can know if a test is unit/integration or e2e.


May be we could simplify into: 
* unit (along the code)
* tests/integration. (mocked system)
* tests/e2e (vm based)


May be we can serparate any future ideas into a different/plan section.

Would this sound more reasonable?

**Comment by mangelajo** [comment]
may be an alternative could be write them with:

* x_integration_test.go (for the mocked integration tests)
* x_test.go (for regular unit tests)
* tests/e2e/xxxx. for the e2e tests (deploymet + VM based) I am working on.
eventually

* tests/benchmark/***

File: test/README.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+Sometimes we need to mock interfaces to make unit testing possible in
+isolation. For that we use the `mockgen` tool from `go.uber.org/mock`
+
+If you want to generate mocks for a package, you can add the reference to
+`/hack/mock.list.txt` and run `make generate` to generate the mocks.
+
+Find more information about using mockgen [here](https://pkg.go.dev/go.uber.org/mock#readme-building-mocks)
+
+
+## Integration testing
```
Comment (romfreiman) [inline test/README.md:62]: Lets elaborate (in few sentences) what is the expected different between integration and e2e? what components will be running in integration? will it also be deployed on kind? or mostly mocked as running binaries? Other differences?

File: test/README.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+should be fast, and should not depend on external services or databases.
+
+We keep the test files for unit tests in the same directory as the code we are
+testing.
+
+i.e.
+
+code in `pkg/log/log.go` should have unit tests in `pkg/log/log_test.go`
+
+We generally use testify, and sometimes Ginkgo/gomega, depending on the nature
```
Comment (mangelajo) [inline test/README.md:42]:
```suggestion
We generally use testify, depending on the nature
```

**Review by hexfusion — APPROVED** [review]
generally fine with the proposal, thanks for putting it up.

File: test/README.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+## Integration testing
+
+Tests the interactions between our different software components, here
+we are not testing dependencies with the operating system or external
+services (beyond the database). External systems or OS interaction
+will be mocked.
+
+Those tests are stored on a separate directory `/test/integration/<component>`,
+i.e. `/test/integration/agent/`, `/test/integration/cmdline/`, etc.
+
```
[the remark this answers is not in the document]
Comment (mangelajo) [reply test/README.md:71]: So , it seems like golang has tools to write coverage data on separate runs, and then merge the coverage data and produce reports out of that:

https://go.dev/doc/build-cover

We would need to investigate how to meld those together between unit and integration, but it's probably doable.

File: test/README.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+## E2E testing
+
+This type of testing verifies the interaction of our software components with
+external software or services, such as the operating system, registries,
+git repositories, etc.
+
+Our stack helps deploy a complete system with `make deploy` and this stack
+should provide everything necessary to perform e2e testing.
+
+We maintain the test files end-to-end tests in `/test/e2e/<topic>`
```
Comment (celdrake) [inline test/README.md:113]:
```suggestion
We maintain the end-to-end test files in `/test/e2e/<topic>`
```

**Review by celdrake — COMMENTED** [review]
Content looks good, just some minor typos / grammar suggestions

File: test/README.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+This project is a complex system with many moving parts. It is important to have a
+coherent testing strategy to ensure that the system works as expected,
+doesn't break over time, and very importantly, make sure that developers know
+where and how to test things.
+
+As testing tools we use two frameworks:
+
+- Ginkgo and Gomega, a BDD-style testing framework for Go, used for more complex
+  testing where BeforeEach/AfterEach is beneficial. It has de benefit of allowing
```
[the remark this answers is not in the document]
Comment (mangelajo) [reply test/README.md:19]: oops

File: test/README.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+We use ginkgo/gomega for these tests, as they are more complex and require
+more setup and teardown than unit tests.
+
+```bash
+make e2e-test
+```
+
+## Command line tool testing
+
+Today we test the command line tool using the bash/github actions today, we
```
[the remark this answers is not in the document]
Comment (mangelajo) [reply test/README.md:140]: thanks 👍

**Comment by mangelajo** [comment]
> Content looks good, just some minor typos / grammar suggestions

Thanks a lot

**Review by celdrake — APPROVED** [review]

**Pull request merged**